diff --git a/src/hotspot/share/compiler/compileBroker.cpp b/src/hotspot/share/compiler/compileBroker.cpp
index f8711e8785a..7a8b5ab3b8f 100644
--- a/src/hotspot/share/compiler/compileBroker.cpp
+++ b/src/hotspot/share/compiler/compileBroker.cpp
@@ -1399,6 +1399,9 @@ nmethod* CompileBroker::compile_method(const methodHandle& method, int osr_bci,
   if (comp == nullptr || compilation_is_prohibited(method, osr_bci, comp_level, directive->ExcludeOption)) {
     return nullptr;
   }
+  if (UseRiftRegions && (comp->is_c2() || comp->is_jvmci())) {
+    return nullptr;
+  }

   if (osr_bci == InvocationEntryBci) {
     // standard compilation
